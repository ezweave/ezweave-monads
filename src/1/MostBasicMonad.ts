interface IMonad<A> {
  bind<B>(callback: (value: A) => Monad<B>): IMonad<B>;
}

abstract class Monad<A> implements IMonad<A> {
  abstract bind<B>(fn: (value: A) => IMonad<B>): IMonad<B>;
}

export class MostBasicMonad<A> extends Monad<A> {
  constructor(private readonly value: A) {
    super();
  }

  public bind<B>(callback: (value: A) => IMonad<B>) {
    return callback(this.value);
  }
}
