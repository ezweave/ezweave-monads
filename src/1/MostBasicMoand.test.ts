import { toUpper } from 'lodash';

import { MostBasicMonad } from './MostBasicMonad';

describe(MostBasicMonad, () => {
  it('can bind', () => {
    const one = new MostBasicMonad('foo');

    const two = one.bind((value) => new MostBasicMonad(toUpper(value)));

    expect(two).toMatchSnapshot();
  });
});
