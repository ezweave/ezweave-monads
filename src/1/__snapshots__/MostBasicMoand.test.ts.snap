// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MostBasicMonad can bind 1`] = `
MostBasicMonad {
  "value": "FOO",
}
`;
